File: src/domain/nutrition.ts
import {
  loadBundledIngredientCatalog,
  type IngredientCatalog,
} from "@/data/ingredientCatalog";
import type {
  QuantifiedIngredient,
  Quantity,
} from "@/domain/quantities";

export interface NutritionFacts {
  readonly calories: number;
  readonly proteinG: number;
  readonly carbohydratesG: number;
  readonly sugarG: number;
  readonly fatG: number;
  readonly fiberG: number;
}

export const EMPTY_NUTRITION_FACTS: NutritionFacts = Object.freeze({
  calories: 0,
  proteinG: 0,
  carbohydratesG: 0,
  sugarG: 0,
  fatG: 0,
  fiberG: 0,
});

export class NutritionValidationError extends Error {
  constructor(message: string) {
    super(message);
    this.name = "NutritionValidationError";
  }
}

export function addNutritionFacts(
  left: NutritionFacts,
  right: NutritionFacts,
): NutritionFacts {
  return Object.freeze({
    calories: left.calories + right.calories,
    proteinG: left.proteinG + right.proteinG,
    carbohydratesG: left.carbohydratesG + right.carbohydratesG,
    sugarG: left.sugarG + right.sugarG,
    fatG: left.fatG + right.fatG,
    fiberG: left.fiberG + right.fiberG,
  });
}

export function scaleNutritionFacts(
  facts: NutritionFacts,
  factor: number,
): NutritionFacts {
  if (!Number.isFinite(factor) || factor < 0) {
    throw new Error("Nutrition scale factor must be finite and non-negative");
  }

  return Object.freeze({
    calories: facts.calories * factor,
    proteinG: facts.proteinG * factor,
    carbohydratesG: facts.carbohydratesG * factor,
    sugarG: facts.sugarG * factor,
    fatG: facts.fatG * factor,
    fiberG: facts.fiberG * factor,
  });
}

export function roundNutritionFacts(
  facts: NutritionFacts,
  digits = 1,
): NutritionFacts {
  if (!Number.isInteger(digits) || digits < 0 || digits > 4) {
    throw new Error("Nutrition rounding digits must be an integer from 0 to 4");
  }

  const factor = 10 ** digits;
  const round = (value: number) => Math.round(value * factor) / factor;

  return Object.freeze({
    calories: round(facts.calories),
    proteinG: round(facts.proteinG),
    carbohydratesG: round(facts.carbohydratesG),
    sugarG: round(facts.sugarG),
    fatG: round(facts.fatG),
    fiberG: round(facts.fiberG),
  });
}

function typicalWeightGrams(
  ingredientId: string,
  catalog: IngredientCatalog,
): number {
  const ingredient = catalog.require(ingredientId);
  const range = ingredient.typicalAmount;

  if (normalizeMassUnit(range.unit) === "g") {
    return (range.minimum + range.maximum) / 2;
  }

  return 100;
}

function normalizeMassUnit(unit: string): "g" | "other" {
  return unit.trim().toLocaleLowerCase("de-DE") === "g" ? "g" : "other";
}

export function quantityToApproximateGrams(
  ingredientId: string,
  quantity: Quantity,
  catalog: IngredientCatalog = loadBundledIngredientCatalog(),
): number {
  switch (quantity.unit) {
    case "g":
      return quantity.amount;
    case "ml":
      return quantity.amount;
    case "piece":
      return quantity.amount * typicalWeightGrams(ingredientId, catalog);
    case "tsp":
      return quantity.amount * 5;
    case "tbsp":
      return quantity.amount * 15;
    case "leaf":
      return quantity.amount;
    case "cube":
      return quantity.amount * 30;
  }
}

export interface NutritionCatalogLike {
  readonly size: number;
  require(ingredientId: string): NutritionFacts;
}

export function calculateNutrition(
  items: readonly QuantifiedIngredient[],
  nutritionCatalog: NutritionCatalogLike,
  ingredientCatalog: IngredientCatalog = loadBundledIngredientCatalog(),
): NutritionFacts {
  let total = EMPTY_NUTRITION_FACTS;

  for (const item of items) {
    const grams = quantityToApproximateGrams(
      item.ingredientId,
      item.quantity,
      ingredientCatalog,
    );
    const per100g = nutritionCatalog.require(item.ingredientId);
    total = addNutritionFacts(
      total,
      scaleNutritionFacts(per100g, grams / 100),
    );
  }

  return total;
}

File: smoothie/nutrition.py
"""Offline nutrition data and aggregation for smoothie quantities."""
from __future__ import annotations

from dataclasses import dataclass
import json
from pathlib import Path

from .ingredient_catalog import IngredientCatalog, load_ingredient_catalog
from .quantities import QuantifiedIngredient, QuantityUnit

DEFAULT_NUTRITION_PATH = Path(__file__).resolve().parents[1] / "data" / "nutrition.json"


@dataclass(frozen=True)
class NutritionFacts:
    calories: float = 0.0
    protein_g: float = 0.0
    carbohydrates_g: float = 0.0
    sugar_g: float = 0.0
    fat_g: float = 0.0
    fiber_g: float = 0.0

    def scaled(self, factor: float) -> "NutritionFacts":
        return NutritionFacts(
            calories=self.calories * factor,
            protein_g=self.protein_g * factor,
            carbohydrates_g=self.carbohydrates_g * factor,
            sugar_g=self.sugar_g * factor,
            fat_g=self.fat_g * factor,
            fiber_g=self.fiber_g * factor,
        )

    def __add__(self, other: "NutritionFacts") -> "NutritionFacts":
        return NutritionFacts(
            calories=self.calories + other.calories,
            protein_g=self.protein_g + other.protein_g,
            carbohydrates_g=self.carbohydrates_g + other.carbohydrates_g,
            sugar_g=self.sugar_g + other.sugar_g,
            fat_g=self.fat_g + other.fat_g,
            fiber_g=self.fiber_g + other.fiber_g,
        )

    def rounded(self, digits: int = 1) -> "NutritionFacts":
        return NutritionFacts(
            calories=round(self.calories, digits),
            protein_g=round(self.protein_g, digits),
            carbohydrates_g=round(self.carbohydrates_g, digits),
            sugar_g=round(self.sugar_g, digits),
            fat_g=round(self.fat_g, digits),
            fiber_g=round(self.fiber_g, digits),
        )

    @classmethod
    def from_dict(cls, data: dict[str, object]) -> "NutritionFacts":
        required = (
            "calories",
            "protein_g",
            "carbohydrates_g",
            "sugar_g",
            "fat_g",
            "fiber_g",
        )
        values: dict[str, float] = {}
        for key in required:
            value = data.get(key)
            if not isinstance(value, (int, float)) or isinstance(value, bool) or value < 0:
                raise ValueError(f"Invalid nutrition value for {key}: {value!r}")
            values[key] = float(value)
        return cls(**values)


class NutritionCatalog:
    def __init__(self, entries: dict[str, NutritionFacts]) -> None:
        self._entries = dict(entries)

    def require(self, ingredient_id: str) -> NutritionFacts:
        try:
            return self._entries[ingredient_id]
        except KeyError as exc:
            raise KeyError(f"Missing nutrition data for ingredient: {ingredient_id}") from exc

    def __len__(self) -> int:
        return len(self._entries)


def load_nutrition_catalog(
    path: Path | str = DEFAULT_NUTRITION_PATH,
    ingredient_catalog: IngredientCatalog | None = None,
) -> NutritionCatalog:
    ingredient_catalog = ingredient_catalog or load_ingredient_catalog()
    try:
        payload = json.loads(Path(path).read_text(encoding="utf-8"))
    except (OSError, json.JSONDecodeError) as exc:
        raise ValueError(f"Could not load nutrition data from {path}: {exc}") from exc
    if not isinstance(payload, dict) or not isinstance(payload.get("ingredients"), dict):
        raise ValueError("Nutrition data must contain an 'ingredients' object")

    raw_entries = payload["ingredients"]
    entries = {ingredient_id: NutritionFacts.from_dict(data) for ingredient_id, data in raw_entries.items()}
    known = {ingredient.id for ingredient in ingredient_catalog.all()}
    unknown = set(entries) - known
    missing = known - set(entries)
    if unknown:
        raise ValueError(f"Nutrition data contains unknown ingredient ids: {sorted(unknown)}")
    if missing:
        raise ValueError(f"Nutrition data missing ingredient ids: {sorted(missing)}")
    return NutritionCatalog(entries)


class NutritionCalculator:
    """Aggregate approximate nutrition without requiring network access."""

    def __init__(self, ingredient_catalog: IngredientCatalog, nutrition_catalog: NutritionCatalog) -> None:
        self.ingredient_catalog = ingredient_catalog
        self.nutrition_catalog = nutrition_catalog

    def calculate(self, items: tuple[QuantifiedIngredient, ...] | list[QuantifiedIngredient]) -> NutritionFacts:
        total = NutritionFacts()
        for item in items:
            grams = self._to_grams(item)
            per_100g = self.nutrition_catalog.require(item.ingredient_id)
            total = total + per_100g.scaled(grams / 100.0)
        return total.rounded()

    def _to_grams(self, item: QuantifiedIngredient) -> float:
        amount = item.quantity.amount
        unit = item.quantity.unit
        ingredient = self.ingredient_catalog.require(item.ingredient_id)
        if unit == QuantityUnit.G:
            return amount
        if unit == QuantityUnit.ML:
            return amount  # Generic density approximation: 1 ml ≈ 1 g.
        if unit == QuantityUnit.PIECE:
            typical = ingredient.typical_amount
            if typical.unit.casefold() == "g":
                return amount * ((typical.minimum + typical.maximum) / 2)
            return amount * 100.0
        if unit == QuantityUnit.TSP:
            return amount * 5.0
        if unit == QuantityUnit.TBSP:
            return amount * 15.0
        if unit == QuantityUnit.LEAF:
            return amount * 1.0
        if unit == QuantityUnit.CUBE:
            return amount * 30.0
        raise ValueError(f"Unsupported nutrition conversion unit: {unit}")
